src/js/program/program-controller.js

```diff
@@ -4,34 +4,60 @@ import cancelable from 'utils/cancelable';
 import { MediaControllerListener } from 'program/program-listeners';
 import Events from 'utils/backbone.events';
 import BackgroundMedia from 'program/background-media';
-import { PLAYER_STATE, STATE_BUFFERING, STATE_IDLE, STATE_PAUSED } from 'events/events';
+import { PLAYER_STATE, STATE_IDLE, STATE_BUFFERING, STATE_PAUSED } from 'events/events';
 import { PlayerError, MSG_CANT_PLAY_VIDEO, ERROR_PLAYLIST_ITEM_MISSING_SOURCE } from 'api/errors';
+import { ItemPromise } from 'controller/next-item-promise';
+import { normalizePlaylistItem } from 'playlist/playlist';
+import Item from 'playlist/item';
 
 class ProgramController extends Events {
     /**
      * ProgramController constructor
      * @param {Model} model - The player's model
      * @param {MediaElementPool} mediaPool - The player's media element pool
+     * @param {PublicApi} publicApi - The player API applied as context in async playlist item promise callback execution
      */
-    constructor(model, mediaPool) {
+    constructor(model, mediaPool, publicApi) {
         super();
 
         this.adPlaying = false;
+        this.apiContext = publicApi;
         this.background = BackgroundMedia();
         this.mediaPool = mediaPool;
         this.mediaController = null;
         this.mediaControllerListener = MediaControllerListener(model, this);
         this.model = model;
         this.providers = new Providers(model.getConfiguration());
-        this.loadPromise = Promise.resolve();
+        this.loadPromise = null;
         this.backgroundLoading = model.get('backgroundLoading');
+        this.itemPromises = [];
 
         if (!this.backgroundLoading) {
             // If background loading is not supported, set the shared media element
             model.set('mediaElement', this.mediaPool.getPrimedElement());
         }
     }
 
+    asyncItem(index, item) {
+        const { model } = this;
+        // Set the player state to buffering if there is a playlist item callback
+        const deferBufferingState = setTimeout(() => {
+            model.set(PLAYER_STATE, STATE_BUFFERING);
+        });
+        return this.getItemPromise(index).run().then((playlistItem) => {
+            clearTimeout(deferBufferingState);
+            if (playlistItem && playlistItem !== item) {
```

Comment:
This block of code runs even if we have already called `replaceItem` in `backgroundLoad` function, because we are checking the result of the `getItemPromise` (not normalized) to `model.get('playlist')[index]` (normalized in `replaceItem`). I will try to come up with a solution before we sync up tomorrow.

Reply:
Thanks @pajong! I moved item replacement in the `ItemPromise` module: https://github.com/jwplayer/jwplayer/pull/3595/commits/dfcf4f3beafc5b084657776af41b3a4625eb9cec

Since it's a new module I will convert it to TypeScript tomorrow and come up with a better name for it.

Now the `ItemPromise.run()` will always return the item that will be used, and if it needed to be normalized and replaced that will already be done. If the callback does not return a promise we'll also resolve the item promise faster.

Reply:
Verified that the preloaded preroll is not destroyed anymore on the VAST side. Thank you! I will run additional tests tomorrow to see if we have anything to refine